components/ui-api-layer/internal/domain/k8s/pod_converter.go

```diff
@@ -0,0 +1,187 @@
+		CreationTimestamp: in.CreationTimestamp.Time,
+		Labels:            in.Labels,
+		Status:            c.podStatusPhaseToGQLStatusType(in.Status.Phase),
+		ContainerStates:   containerStates,
+		JSON:              gqlJSON,
+	}, nil
+}
+
+func (c *podConverter) ToGQLs(in []*v1.Pod) ([]gqlschema.Pod, error) {
+	var result []gqlschema.Pod
+	for _, u := range in {
+		converted, err := c.ToGQL(u)
+		if err != nil {
+			return nil, err
+		}
+
+		if converted != nil {
+			result = append(result, *converted)
+		}
+	}
+	return result, nil
+}
+
+func (c *podConverter) containerStatusesToGQLContainerStates(in []v1.ContainerStatus) []gqlschema.ContainerState {
+	containerStates := []gqlschema.ContainerState{}
+
+	for _, containerStatus := range in {
+		if containerStatus.State.Waiting != nil {
+			containerStates = append(containerStates, c.getWaitingContainerState(containerStatus.State.Waiting))
+		} else if containerStatus.State.Terminated != nil {
+			containerStates = append(containerStates, c.getTerminatedContainerState(containerStatus.State.Terminated))
+		} else if containerStatus.State.Running != nil {
+			containerStates = append(containerStates, c.getRunningContainerState())
+		} else {
+			containerStates = append(containerStates, c.getDefaultContainerState())
+		}
+	}
+
+	return containerStates
+}
+
+func (c *podConverter) getWaitingContainerState(in *v1.ContainerStateWaiting) gqlschema.ContainerState {
+	if in == nil {
+		return gqlschema.ContainerState{
+			State: gqlschema.ContainerStateTypeWaiting,
+		}
+	}
+
+	var reason, message *string
+
+	if in.Reason != "" {
+		tmp := in.Reason
+		reason = &tmp
+	}
+
+	if in.Message != "" {
+		tmp := in.Message
+		message = &tmp
+	}
+
+	return gqlschema.ContainerState{
+		State:   gqlschema.ContainerStateTypeWaiting,
+		Reason:  reason,
+		Message: message,
+	}
+}
+
+func (c *podConverter) getRunningContainerState() gqlschema.ContainerState {
+	return gqlschema.ContainerState{
+		State:   gqlschema.ContainerStateTypeRunning,
+		Reason:  nil,
+		Message: nil,
+	}
+}
+
+func (c *podConverter) getTerminatedContainerState(in *v1.ContainerStateTerminated) gqlschema.ContainerState {
+	if in == nil {
+		return gqlschema.ContainerState{
+			State: gqlschema.ContainerStateTypeTerminated,
+		}
+	}
+
+	var reason, message *string
+
+	if in.Reason != "" {
+		tmp := in.Reason
+		reason = &tmp
+	} else if in.Signal != 0 {
+		tmp := fmt.Sprintf("Signal: %d", in.Signal)
+		reason = &tmp
+	} else {
+		tmp := fmt.Sprintf("Exit code: %d", in.ExitCode)
+		reason = &tmp
+	}
+
+	if in.Message != "" {
+		tmp := in.Message
+		message = &tmp
+	}
+
+	return gqlschema.ContainerState{
+		State:   gqlschema.ContainerStateTypeTerminated,
+		Reason:  reason,
+		Message: message,
+	}
+}
+
+func (c *podConverter) getDefaultContainerState() gqlschema.ContainerState {
+	return c.getWaitingContainerState(nil)
+}
+
+func (c *podConverter) podToGQLJSON(in *v1.Pod) (gqlschema.JSON, error) {
```

Comment:
Probably name `podSpecificToGQLJSON` would be better.

Reply:
Wouldn't that suggest only the PodSpec field of Pod is being converted while we want to convert whole Pod?